File: wagerproof-ios-native/Wagerproof/Features/Agents/Components/AgentColorPalette.swift
import SwiftUI
import UIKit

/// Helpers for the per-agent `avatar_color` field, which is either a hex
/// string ("#6366f1") or a gradient pair ("gradient:#6366f1,#ec4899"). Ports
/// the `parseAvatarColor` / `getPrimaryColor` / `getSecondaryColor` helpers
/// duplicated across multiple RN components.
enum AgentColorPalette {
    static func primary(for raw: String) -> Color {
        Color(hexString: primaryHex(for: raw)) ?? Color(hex: 0x6366F1)
    }

    static func secondary(for raw: String) -> Color {
        Color(hexString: secondaryHex(for: raw)) ?? primary(for: raw)
    }

    static func gradient(for raw: String) -> (Color, Color) {
        (primary(for: raw), secondary(for: raw))
    }

    /// A two-color gradient for avatar tiles that is *always* visibly two-tone.
    /// When `avatar_color` is a solid hex (primary == secondary), `[primary,
    /// secondary]` would render a flat fill — so we derive a darker partner to
    /// give the tile real top-left→bottom-right depth. Gradient-pair colors are
    /// used as-is.
    static func avatarGradient(for raw: String) -> [Color] {
        let p = primary(for: raw)
        if secondaryHex(for: raw).caseInsensitiveCompare(primaryHex(for: raw)) == .orderedSame {
            return [p, p.shaded(by: 0.55)]
        }
        return [p, secondary(for: raw)]
    }

    static func primaryHex(for raw: String) -> String {
        if raw.hasPrefix("gradient:") {
            let stripped = String(raw.dropFirst("gradient:".count))
            return stripped.split(separator: ",").first.map(String.init) ?? "#6366f1"
        }
        return raw
    }

    static func secondaryHex(for raw: String) -> String {
        if raw.hasPrefix("gradient:") {
            let stripped = String(raw.dropFirst("gradient:".count))
            let parts = stripped.split(separator: ",")
            if parts.count >= 2 { return String(parts[1]) }
            return parts.first.map(String.init) ?? "#6366f1"
        }
        return raw
    }
}

extension Color {
    /// Multiply RGB by `factor` (<1 darkens, >1 lightens), clamped to [0,1].
    /// Used to derive a gradient partner for solid avatar colors.
    func shaded(by factor: Double) -> Color {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        guard UIColor(self).getRed(&r, green: &g, blue: &b, alpha: &a) else { return self }
        func cl(_ v: CGFloat) -> Double { Double(min(max(v * CGFloat(factor), 0), 1)) }
        return Color(.sRGB, red: cl(r), green: cl(g), blue: cl(b), opacity: Double(a))
    }

    /// Parses "#RRGGBB" or "#AARRGGBB" / "RRGGBB" strings. Returns nil on
    /// malformed input.
    init?(hexString: String) {
        var s = hexString.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6 || s.count == 8 else { return nil }
        var value: UInt64 = 0
        guard Scanner(string: s).scanHexInt64(&value) else { return nil }
        let r, g, b, a: Double
        if s.count == 6 {
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
            a = 1.0
        } else {
            a = Double((value >> 24) & 0xFF) / 255
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
        }
        self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/AgentColorPalette.kt
package com.wagerproof.app.features.agents

import androidx.compose.ui.graphics.Color

/**
 * Helpers for the per-agent `avatar_color` field, which is either a hex string
 * ("#6366f1") or a gradient pair ("gradient:#6366f1,#ec4899"). Port of iOS
 * `AgentColorPalette.swift`. Shared by every agents-local card/hero/sheet.
 */
object AgentColorPalette {
    /** Fallback indigo when a color string is missing/malformed (Swift 0x6366F1). */
    val fallback = Color(0xFF6366F1)

    fun primary(raw: String): Color = colorFromHexString(primaryHex(raw)) ?: fallback

    fun secondary(raw: String): Color = colorFromHexString(secondaryHex(raw)) ?: primary(raw)

    fun gradient(raw: String): Pair<Color, Color> = primary(raw) to secondary(raw)

    /**
     * A two-color gradient for avatar tiles that is ALWAYS visibly two-tone.
     * When `avatar_color` is a solid hex (primary == secondary), derive a darker
     * partner so the tile has real top-left→bottom-right depth. Gradient pairs
     * are used as-is.
     */
    fun avatarGradient(raw: String): List<Color> {
        val p = primary(raw)
        return if (secondaryHex(raw).equals(primaryHex(raw), ignoreCase = true)) {
            listOf(p, p.shaded(0.55))
        } else {
            listOf(p, secondary(raw))
        }
    }

    fun primaryHex(raw: String): String {
        if (raw.startsWith("gradient:")) {
            val stripped = raw.removePrefix("gradient:")
            return stripped.split(",").firstOrNull()?.trim() ?: "#6366f1"
        }
        return raw
    }

    fun secondaryHex(raw: String): String {
        if (raw.startsWith("gradient:")) {
            val stripped = raw.removePrefix("gradient:")
            val parts = stripped.split(",")
            if (parts.size >= 2) return parts[1].trim()
            return parts.firstOrNull()?.trim() ?: "#6366f1"
        }
        return raw
    }
}

/** Multiply RGB by [factor] (<1 darkens, >1 lightens), clamped to [0,1]. */
fun Color.shaded(factor: Double): Color {
    fun cl(v: Float): Float = (v * factor.toFloat()).coerceIn(0f, 1f)
    return Color(red = cl(red), green = cl(green), blue = cl(blue), alpha = alpha)
}

/**
 * Parses "#RRGGBB" / "RRGGBB" / "#AARRGGBB" strings. Returns null on malformed
 * input (matches Swift `Color(hexString:)`).
 */
fun colorFromHexString(hexString: String): Color? {
    var s = hexString.trim()
    if (s.startsWith("#")) s = s.substring(1)
    if (s.length != 6 && s.length != 8) return null
    val value = s.toULongOrNull(16) ?: return null
    return if (s.length == 6) {
        Color(
            red = ((value shr 16) and 0xFFu).toInt() / 255f,
            green = ((value shr 8) and 0xFFu).toInt() / 255f,
            blue = (value and 0xFFu).toInt() / 255f,
            alpha = 1f,
        )
    } else {
        Color(
            alpha = ((value shr 24) and 0xFFu).toInt() / 255f,
            red = ((value shr 16) and 0xFFu).toInt() / 255f,
            green = ((value shr 8) and 0xFFu).toInt() / 255f,
            blue = (value and 0xFFu).toInt() / 255f,
        )
    }
}
